### PR TITLE
 ENH: linalg: memoize get_lapack/blas_funcs to speed it up 

### PR DESCRIPTION
**Pull request by pv**
The `scipy.linalg.get_lapack/blas_funcs` are relatively slow and can take time comparable to the Fortran codes for small matrices.

Speed it up (~10x) by memoization: the fortran routine for a given dtype never changes.

This is prompted by gh-10512 --- after this PR, `get_lapack_funcs` runtime goes to ~0.5µs for cache hit, whereas `cholesky(x)` for a 1x1 matrix takes 16µs, so getting the func becomes insignificant.

Benchmark:
```
       before           after         ratio
     [2331b37b]       [fdf08144]
     <master>         <memoize-get-funcs>
-     22.9±0.03μs       15.8±0.1μs     0.69  linalg.GetFuncs.time_small_cholesky
-     8.24±0.05μs      1.20±0.01μs     0.15  linalg.GetFuncs.time_get_blas_funcs_2
-     5.73±0.09μs         554±20ns     0.10  linalg.GetFuncs.time_get_blas_funcs
```

### CODE REVIEW

**Review by larsoner — COMMENTED** [review]
Other than one generalization idea that I'm not sure applies to LAPACK funcs properly but might, looks good to me. This is something I just thought about when doing some code optimization last week, this would be very useful!

The other thing I have found is that for small transformations, ~1/3 of the time can be spent in `get_lapack_funcs`, ~1/3 of the time can be spent in `_compute_lwork`, and ~1/3 of the time is in actual computation. Do you think it would be feasible to implement something similar for `_compute_lwork`? I don't know enough about what `_compute_lwork` actually depends on to know.

File: scipy/linalg/blas.py [hunk 1]
```diff
@@ -361,6 +362,39 @@ def _get_funcs(names, arrays, dtype,
...
+    memo = {}
+    func.memo = memo
+
+    @functools.wraps(func)
+    def getter(names, arrays=(), dtype=None):
+        if arrays == ():
+            key = (names, dtype)
+            value = memo.get(key)
+        elif len(arrays) == 1 and dtype is None:
+            key = (names, arrays[0].dtype, arrays[0].flags.fortran)
```
[the remark this answers is not in the document]
Comment (pv) [reply scipy/linalg/blas.py]: Yes, changed. It's 10% slower but that's probably not important.

**Comment by pv** [comment]
Not sure if `_compute_lwork` can be memoized --- at least as simply (you need to limit cache size, since there are many different array sizes). Maybe worth thinking about later on separately.

**Comment by pv** [comment]
OTOH, 8d717dd393dc6c6da made some strange changes to `_compute_lwork` that probably slow it down...

**Comment by larsoner** [comment]
Agreed, we can try to speed up `_compute_lwork` separately someday. This is already a nice improvement.

**Review by larsoner — APPROVED** [review]

**Comment by ilayn** [comment]
Ah nice. +1 from me too hence clicking the button.

**Pull request merged**